### PR TITLE
ci/cd: build linux-musl artifacts

### PR DESCRIPTION
**Pull request by avengermsoft**
Hey!

This is an attempt to add linux-musl support to the project.

### CODE REVIEW

**Comment by avengermsoft** [comment]
Jackett on OpenWrt is work!

```
root@RPi4:~# uname -a
Linux RPi4 5.10.82 #0 SMP PREEMPT Sun Nov 28 15:32:45 2021 aarch64 GNU/Linux
root@RPi4:~# ldd
musl libc (aarch64)
Version 1.2.2
Dynamic Program Loader
Usage: ldd [options] [--] pathname
root@RPi4:~#
```

```
2022-01-29 13:35:18.8808 Info Starting Jackett v0.0.0 
2022-01-29 13:35:20.7705 Info Environment version: 6.0.1 (/tmp/publish/) 
2022-01-29 13:35:20.7705 Info OS version: Unix 5.10.82.0 (64bit OS) (64bit process) 
2022-01-29 13:35:20.7705 Info Jackett variant: CoreLinuxMuslArm64 
2022-01-29 13:35:20.7705 Info Running in Docker: No 
2022-01-29 13:35:20.7705 Info ThreadPool MaxThreads: 32767 workerThreads, 1000 completionPortThreads 
2022-01-29 13:35:20.7705 Info App config/log directory: /opt/Jackett 
2022-01-29 13:35:20.7805 Info Using proxy: Disabled 
2022-01-29 13:35:20.7805 Info Using FlareSolverr: No 
2022-01-29 13:35:20.7805 Info Using HTTP Client: HttpWebClient2 
2022-01-29 13:35:20.8134 Info Loading Native indexers ... 
2022-01-29 13:35:21.3078 Info Loaded 86 Native indexers: abnormal, alpharatio, anidex, anidub, animebytes, Animedia, animetorrents, anthelion, avistaz, bakabt, bb, beyond-hd-api, bithdtv, bjshare, broadcasthenet, brokenstones, cgpeers, cinecalidad, cinemaz, danishbytes, darmowetorenty, divxtotal, dontorrent, epublibre, erai-raws, exoticaz, filelist, funfile, fuzer, gazellegames, gimmepeers, hdbitsapi, hdspace, hdtorrents, hebits, immortalseed, iptorrents, libble, lostfilm, magnetico, mejortorrent, morethantv, myanonamouse, ncore, nebulance, newpct, norbits, oppaitime, orpheus, partis, passthepopcorn, piratethenet, pixelhd, polishtracker, pornolab, pretome, privatehd, rarbg, redacted, retroflix, revolutiontt, rutracker, scenehd, scenetime, seals, secretcinema, sharewoodapi, shazbat, ShizaProject, speedapp, speedcd, subsplease, toloka, torrentbytes, torrentday, torrentech, torrentheaven, torrentmafya, torrentnetwork, torrentscsv, torrentsyndikat, tvstore, xspeeds, xthor-api, zonaq, animetosho 
2022-01-29 13:35:21.3186 Info Loading Cardigann indexers from: /opt/cardigann/definitions/, /etc/xdg/cardigan/definitions/, /tmp/publish/Definitions 
2022-01-29 13:35:27.3617 Info Loaded 491 Cardigann indexers: ztracker, zooqle, zomb, zetorrents, zelkaorg, zamundanet, yts, yourbittorrent, yggtorrent, yggcookie, ydypt, xxxtorrents, xxxtor, xxxadulttorrent, xwtorrents, xwtclassics, xbytesv2, xiteme, woot, wihd, whiteangel, wdt, vsttorrents, vsthouse, vizuk, unlimitz, uniongang, unionfansub, uniondht, underversel, underverse, uhdbits, uhd-heaven, u2, twilightszoom, twilight, tvroad, tvchaosuk, turktorrent, turkseed, trupornolabs, tribalmixes, trezzor, trancetraffic, trackeros-api, totheglorycookie, totheglory, totallykids, torrentz2eu, torrentwhiz, torrentview, torrentv, torrentsurf, torrentslocal, torrentsir, torrentseeds-api, torrentsectorcrew, torrentqq, torrentproject2, torrentparadise, torrentoyunindir, torrentmax, torrentmasters, torrentlt, torrentleech, torrentleech-pl, torrentland, torrentkitty, torrenting, torrenthr, torrentgalaxy, torrentfunk, torrentdownloads, torrentdownload, torrentdb, torrentccf, torrentby, torrentbd, torrent9clone, torrent9, torrent-turk, torrent-pirat, torrent-explosiv, toros, torlook, torlock, tokyotosho, tlfbits, tjupt, thevault, theshow, theshinning, thesceneplace, theplace, thepiratebay, theoccult, themixingbowl, theleachzone, thehorrorcharnel, thegeeks, thefallingangels, theempire, theaudioscene, the-devils-lounge, teracod, tellytorrent-api, tekno3d, teamos, teamhd, tasmanit, tapochek, swarmazon, superbits, sukebeinyaasi, sugoimusic, springsunday, sportscult, sporthd, spiritofrevolution, spidertk, speedtorrentreloaded, speedmasterhd, soundpark, soulvoice, sosulki, solidtorrents, snowpt, slosoul, sktorrent, sktorrent-org, skipthecommercials-api, siambit, showrss, sharewood, shareisland-api, sexypics, sexxi, selezen, seedfile, sdbits, scenerush, scenepalace, rutracker-ru, rutor, rustorka, rus-media, rptorrents, romanianmetaltorrents, RockBox, riperam, rintornet, rintor, rgfootball, resurrectthenet, reelflix-api, redstartorrent, redbits, rapidzona, rainbowtracker, racingforme, racing4everyone, r3vwtf, pwtorrents, puurhollands, putao, pussytorrents, purovicio, puntotorrent, punkshorror, pttime, ptsbao, ptmsg, pthome, ptfiles, pterclub, prostylex, proporno, proaudiotorrents, potuk, portugas, pornotorrent, pornotor, pornoslab, pornolive, pornleech, pornforall, pornbits, pornbay, polishsource, pixelcove, pirateiro, piratbitl, piratbit, piratadigital, pier720, peersfm, peerjunkies, peeratiko, pctorrent, parnuxi, p2pelite, p2pbg, oxtorrent, ourbits, oshenpt, opensharing, opencd, onejav, omgwtftrackr, nyaasi, noname-clubl, noname-club, nntt, nitro, nipponsei, newstudiol, newstudio, newretro, newpct-me, netlab, nethd, netcosmo, nbtorrents, myspleen, mypornclub, mvgroupmain, mvgroupforum, muziekfabriek, mteamtp2fa, mteamtp, movietorrentz, movietorrent, moviesite, moviesdvdr, mousebits, mononokebt, mnv, mma-torrents, mixtapetorrent, mircrew, milkie, mikan, microbit, metaltracker, mesevilag, megapeer, megamixtracker, mediamaatjes, marinetracker, majomparade, magico, magicheaven, madsrevolution, mactorrentsdownload, losslessclub, linuxtracker, linkomanija, limetorrentsclone, limetorrents, libranet, lesaloon, lepornoinfo, lemonhd, legittorrents, leech24, learnflakes, learnbits, lat-team, lastfiles, krazyzone, kisssub, kinozal, kinorun, kinonavse100, kickasstorrents-ws, kickasstorrents-to, keepfriends, karagarga, jptv, jpopsuki, iv-torrents, itorrent, isohunt2, internetarchive, insanetracker, indietorrents, immortuos, ilcorsaroverde, ilcorsaronero, ilcorsaroblu, idope, ibit, huntorrent, hqsource, hellastz, hdzone, hdu, hdturk, hdtorrentsit, hdtime, hdspain, hdsky, hdroute, hdonly, hdolimpo-api, hdmonkey, hdme, hdhouse, hdhome, hdforever, hddolby, hdcztorrent, hdcity, hdchina, hdcenter, hdc, hdbitscom, hdbits, hdatmos, hdarea, hdai, hd4fans, hawke-uno, happyfappy, hamsterstudio, haitang, haidan, gtorrentpro, greekteam, greekdiamond, glodls, gktorrent, girotorrent, gigatorrents, gfxpeers, genesismovement, generationfree, gaytorrentru, gay-torrentsorg, gay-torrents, gamestorrents, funkytorrents, frozenlayer, freshmeat, fouducinema, firebit, finvip, finelite, film-paleis, filelisting, femdomcult, fantastiko, fantasticheaven, fanoin, falkonvisionteam, eztv, exttorrents, extremlymtorrents, extremetorrents, extremebits, extratorrent-st, exkinoray, ettv, estone, esharenet, eniahd, empornium2fa, empornium, elitetorrent-biz, ehentai, ebooks-shares, ebookparadijs, dxp, dragonworldreloaded, docspedia, dmhy, divteam, discfan, dimeadozen, digitalcore, diablotorrent, desitorrents-api, demonoid, deildu, datatalli, datascene-api, darktracker, dark-shadow, dariustracker, cryptichaven, crt2fa, crnaberza, crazyspirits, crazyhd, cpasbienclone, cpasbien, concertos, concen, comicat, coastalcrew, classix, cinematik, cinemamovies, cinemageddon, chilebt, chdbits, ceskeforum, ccfbits, catorrent, cathoderaytube, casatorrent, cartoonchaos, carphunter, carpathians, byrutor, byrbt, bwtorrents, btsow, btschool, btnext, btetree, btdigg, bt4g, brsociety-api, broadcity, brasiltracker, boxingtorrents, borgzelle, bootytape, booktracker, blutopia-api, bluesbrothers, bluebirdhd, bitturk, Bittorrentfiles, bitspyder, bitsexy, bitsearch, bitru, bithumen, bithorlo, bitded, bitcityreloaded, bitbazis, bit-titan, bigfangroup, bibliotik, beyond-hd, beyond-hd-oneurl, bestcore, beitai, bangumi-moe, baibako, badasstorrents, backups, aussierules, audionews, audiobookbay, asiancinema, arenabg, arabp2p, arabafenice, anisource, aniRena, animeworld-api, animelayer, animeclipse, anime-free, anilibria, anaschcc, amigosshare, alleenretail, aither-api, aidoruonline, aftershock, acrossthetasman, acidlounge, acgsou, acgrip, abtorrents, 7torrents, 52pt, 4thd, 3dtorrents, 3changtrai, 2fast4you, 24rolika, 1ptbar, 1337x, 0magnet, 0daykiev 
2022-01-29 13:35:27.3617 Info Loaded 577 indexers in total 
2022-01-29 13:35:27.3617 Info Adding aggregate indexer ('all' indexer) ... 
2022-01-29 13:35:27.3864 Info Adding filter indexer ('type:public' indexer) ... 
2022-01-29 13:35:27.3864 Info Adding filter indexer ('type:private' indexer) ... 
2022-01-29 13:35:27.3864 Info Adding filter indexer ('type:semi-public' indexer) ... 
2022-01-29 13:35:27.9181 Info Jackett startup finished in 10.352 s 
```

**Comment by avengermsoft** [comment]
I had to build a library for OpenWrt with a patch: 

```
--- a\gcc\common.mk	Thu Jan 27 15:38:48 2022
+++ b\gcc\common.mk	Sat Jan 29 00:40:47 2022
@@ -115,7 +115,7 @@
 		--disable-decimal-float \
 		--with-diagnostics-color=auto-if-env \
 		--enable-__cxa_atexit \
-		--disable-libstdcxx-dual-abi \
+		--enable-libstdcxx-dual-abi \
 		--with-default-libstdcxx-abi=new
 ifneq ($(CONFIG_mips)$(CONFIG_mipsel),)
   GCC_CONFIGURE += --with-mips-plt
```
Otherwise I got errors 
```
Error relocating /tmp/publish/jackett: _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEED2Ev: symbol not found
Error relocating /tmp/publish/jackett: _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE9_M_createERmm: symbol not found
Error relocating /tmp/publish/jackett: _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE9_M_assignERKS4_: symbol not found
Error relocating /tmp/publish/jackett: _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE10_M_replaceEmmPKcm: symbol not found
Error relocating /tmp/publish/jackett: _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE9_M_mutateEmmPKcm: symbol not found
Error relocating /tmp/publish/jackett: _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE9_M_appendEPKcm: symbol not found
Error relocating /tmp/publish/jackett: _ZNKSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE5rfindEPKcmm: symbol not found
Error relocating /tmp/publish/jackett: _ZNKSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE5rfindEcm: symbol not found
Error relocating /tmp/publish/jackett: _ZNKSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE12find_last_ofEPKcmm: symbol not found
Error relocating /tmp/publish/jackett: _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE8_M_eraseEmm: symbol not found
Error relocating /tmp/publish/jackett: _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE14_M_replace_auxEmmmc: symbol not found
Error relocating /tmp/publish/jackett: _ZNKSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE4findEcm: symbol not found
Error relocating /tmp/publish/jackett: _ZNSt7__cxx1118basic_stringstreamIcSt11char_traitsIcESaIcEEC1ERKNS_12basic_stringIcS2_S3_EESt13_Ios_Openmode: symbol not found
Error relocating /tmp/publish/jackett: _ZSt7getlineIcSt11char_traitsIcESaIcEERSt13basic_istreamIT_T0_ES7_RNSt7__cxx1112basic_stringIS4_S5_T1_EES4_: symbol not found
Error relocating /tmp/publish/jackett: _ZNKSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE17find_first_not_ofEPKcmm: symbol not found
Error relocating /tmp/publish/jackett: _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE7reserveEm: symbol not found
Error relocating /tmp/publish/jackett: _ZNSt7__cxx1118basic_stringstreamIcSt11char_traitsIcESaIcEEC1ESt13_Ios_Openmode: symbol not found
Error relocating /tmp/publish/jackett: _ZNKSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE4findEPKcmm: symbol not found
Error relocating /tmp/publish/jackett: _ZNKSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE7compareEmmRKS4_: symbol not found
Error relocating /tmp/publish/jackett: _ZNKSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE7compareEmmPKc: symbol not found
Error relocating /tmp/publish/jackett: _ZTTNSt7__cxx1118basic_stringstreamIcSt11char_traitsIcESaIcEEE: symbol not found
Error relocating /tmp/publish/jackett: _ZTVNSt7__cxx1115basic_stringbufIcSt11char_traitsIcESaIcEEE: symbol not found
```

File: src/Jackett.Updater/Program.cs [hunk 1]
```diff
@@ -478,7 +478,6 @@ private void ProcessUpdate(UpdaterConsoleOptions options)
                 "Definitions/tntfork.yml",
                 "Definitions/tntvillage.yml",
                 "Definitions/topnow.yml",
-                "Definitions/torrent-paradise-ml.yml",
```
Comment (ngosang) [inline src/Jackett.Updater/Program.cs:481]: should be in another pr

**Review by ngosang — APPROVED** [review]
I think it is right. Maybe you have to update the readme/documentation to inform about all the OS we support.

**Comment by avengermsoft** [comment]
I want to add native OpenWrt support. Currently there are problems with the libstdc++ library. [PR on OpenWrt](https://github.com/openwrt/openwrt/pull/5014)

**Comment by avengermsoft** [comment]
There is no pgrep command with the -u option in the Jackett_launcher.sh script. There is an analogue - the pgrep-procps-ng command. Is it possible somehow to use two commands in this script?

```
Wed Feb  2 11:21:51 2022 daemon.info jackett_launcher.sh[20670]: debug whoami = jackett
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]: pgrep: unrecognized option: u
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]: BusyBox v1.34.1 (2021-11-27 20:49:10 UTC) multi-call binary.
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]:
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]: Usage: pgrep [-flanovx] [-s SID|-P PPID|PATTERN]
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]:
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]: Display process(es) selected by regex PATTERN
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]:
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]: 	-l	Show command name too
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]: 	-a	Show command line too
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]: 	-f	Match against entire command line
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]: 	-n	Show the newest process only
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]: 	-o	Show the oldest process only
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]: 	-v	Negate the match
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]: 	-x	Match whole name (not substring)
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]: 	-s	Match session ID (0 for current)
Wed Feb  2 11:21:51 2022 daemon.err jackett_launcher.sh[20670]: 	-P	Match parent process ID
```

**Comment by avengermsoft** [comment]
To run Jackett on Openwrt, do the following:
1. Compile OpenWrt with a [patch](https://github.com/openwrt/openwrt/pull/5014/files) for your platform. Requires only one library `libstdcpp6_*.ipk`
2. Compile Jackett for your platform `dotnet publish Jackett.Server --configuration Release --runtime linux-musl-(x64|arm|arm64) --framework net6.0 --self-contained`.
3. Install packages `opkg update && opkg install icu70 icu-full-data70 libstdcpp6 libintl-full8 procps-ng procps-ng-pgrep coreutils-whoami` to your router
4. Copy Jackett to your router.
5. Create a `netcoredeps` subdirectory under the Jackett directory and copy the `libstdc++.so.*` file from the `libstdcpp6_*.ipk` package compiled in step 1 into it.
6. Run Jackett ./jackett

**Comment by ngosang** [comment]
> There is no pgrep command with the -u option in the Jackett_launcher.sh script. There is an analogue - the pgrep-procps-ng command. Is it possible somehow to use two commands in this script?

I'm not sure about OpenWRT but the BusyBox included in Alpine Linux has the command and it's working. Could you double check? I prefer to keep the file simple. Another fix could be creating a symbolic link `pgrep => pgrep-procps-ng` in your install script.
```
pgrep
BusyBox v1.34.1 (2021-11-23 00:57:35 UTC) multi-call binary.

Usage: pgrep [-flanovx] [-s SID|-P PPID|PATTERN]

Display process(es) selected by regex PATTERN

        -l      Show command name too
        -a      Show command line too
        -f      Match against entire command line
        -n      Show the newest process only
        -o      Show the oldest process only
        -v      Negate the match
        -x      Match whole name (not substring)
        -s      Match session ID (0 for current)
        -P      Match parent process ID
        -u EUID Match against effective UID
        -U UID  Match against UID
/ # ps aux
PID   USER     TIME  COMMAND
    1 root      0:17 /sbin/tini -- docker-entrypoint.sh gerbera --port 49494 --config /var/run/gerbera/config.xml
    9 gerbera   8:32 gerbera --port 49494 --config /var/run/gerbera/config.xml
/ # pgrep -u gerbera gerbera
9
```

You have to add another task in Azure to include the file `install_service_openwrt.sh` in musl artifacts. Take a look => https://github.com/Jackett/Jackett/blob/d22820080fe6a7957c7221a1e8aafad061a77dc1/azure-pipelines.yml#L209

Except for those 2 things the PR looks good to me.

**Comment by lincolnthalles** [comment]
This is PR working fine on Alpine Linux. Please merge.

JackettUpdater won't work until Musl is officially supported. Openrc service will require a new 'install_service_openrc.sh' or manual setup.

After building this PR and pushing Jackett.Binaries.LinuxMuslAMDx64.tar.gz into container, I managed to get Jackett working this way:

```sh
# Make sure you are running alpine edge
cat > /etc/apk/repositories << EOF; $(echo)
http://dl-cdn.alpinelinux.org/alpine/edge/main/
http://dl-cdn.alpinelinux.org/alpine/edge/community/
#http://dl-cdn.alpinelinux.org/alpine/edge/testing/
EOF
apk update && apk add --no-cache --upgrade apk-tools
apk upgrade --available && apk cache --purge
sync && reboot

# requirements on alpine/musl:
apk add --no-cache sqlite-libs gcompat bash icu-libs krb5-libs libgcc libintl libssl1.1 libstdc++ zlib

# Extract Musl build to /opt:
cd /opt
tar -C . -xvf "Jackett.Binaries.LinuxMuslAMDx64.tar.gz"
rm "Jackett.Binaries.LinuxMuslAMDx64.tar.gz"

# In case Jackett was built on Windows
chmod +x /opt/Jackett/jackett
# updater is not yet available
chmod +x /opt/Jackett/JackettUpdater

# setup user and group
addgroup -S jackett
adduser -S -D -h /opt/JackettData -s /sbin/nologin jackett -G jackett
addgroup -S media && addgroup -S jackett media

# setup permissions
chown -R jackett:jackett /opt/Jackett
chown -R jackett:jackett /opt/JackettData

# create openrc service
tee /etc/init.d/jackett <<EOF
#!/sbin/openrc-run

command="/opt/Jackett/jackett"
command_args="--NoRestart"
command_user="jackett:jackett"
command_background=true

name="\${RC_SVCNAME}"
basedir="/opt/Jackett"
datadir="/opt/JackettData"

start_stop_daemon_args="--chdir \${basedir}"
pidfile="/run/\${RC_SVCNAME}.pid"  

depend() {
  after network-online
  use network-online 
}

stop_pre() {
  # from jackett_launcher.sh
  while pgrep -u "\$(whoami)" JackettUpdater > /dev/null; do
    sleep 1
  done
}

start_pre() {
  checkpath -q -d -m 0777 -o "\${command_user}" \\
  \${basedir} \\
  \${datadir}
  export XDG_CONFIG_HOME=\${datadir}
  rm -f "\${datadir}/\${RC_SVCNAME}.pid"
}
EOF

chmod +x /etc/init.d/jackett

# add service to autostart
rc-update add jackett default
rc-service jackett start

# debugging
rc-service jackett status
cat /var/log/messages
cat /opt/JackettData/Jackett/log.txt
```

**Comment by avengermsoft** [comment]
https://github.com/openwrt/openwrt/commit/3c06a344e9c7c03c49c9153342e68a5390651323 merged.

**Comment by ngosang** [comment]
@garfield69 @ilike2burnthing I think this is good enough. I'm going to publish a new release to test the updater. The documentation is here:

- [Alpine Linux](https://github.com/Jackett/Jackett/wiki/Installation-on-Alpine-Linux)
- [OpenWrt](https://github.com/Jackett/Jackett/wiki/Installation-on-OpenWrt)
